packages/cli/src/models/compiler/solidity/CompilerProvider.ts

```diff
@@ -0,0 +1,212 @@
+  public _version: string;
+
+  constructor(version: string) {
+    this._version = version;
+  }
+
+  public version(): string {
+    return this._version;
+  }
+
+  public async compile(input: any): Promise<solc.CompilerOutput> {
+    const output = child.execSync('solc --standard-json', { input: JSON.stringify(input), env: SOLC_BIN_ENV });
+    return JSON.parse(output.toString());
+  }
+}
+
+export async function resolveCompilerVersion(requiredSemver: string | string[]): Promise<SolcBuild> {
+  // Create an array with all unique semver restrictions (dropping initial 'v' if set manually by the user)
+  const requiredSemvers = uniq(compact(castArray(requiredSemver)))
+    .map(str => str.startsWith('v') ? str.slice(1) : str);
+
+  // TODO: Pin the compiler release, so a new release does not cause all contracts to be recompiled and thus redeployed
+  const solcList = await getAvailableCompilerVersions();
+  const build = await getCompilerVersion(requiredSemvers, solcList);
+  return build;
+}
+
+export async function fetchCompiler(build: SolcBuild): Promise<SolcCompiler> {
+  // Try local compiler and see if version matches
+  const localVersion = await localCompilerVersion();
+  if (localVersion && compilerVersionsMatch(localVersion, build.longVersion)) {
+    log.info(`Using local solc compiler found`);
+    return new SolcBinCompiler(localVersion);
+  }
+
+  // Go with emscriptem version if not
+  const localFile = path.join(SOLC_CACHE_PATH, build.path);
+  if (!fs.existsSync(localFile)) await downloadCompiler(build, localFile);
+  const compilerBinary = getCompilerBinary(localFile);
+
+  // Wrap emscriptem with solc-wrapper
+  return new SolcjsCompiler(compilerBinary);
+}
+
+export async function getCompiler(requiredSemver: string | string[]): Promise<SolcCompiler> {
+  const version = await resolveCompilerVersion(requiredSemver);
+  return fetchCompiler(version);
+}
+
+async function localCompilerVersion(): Promise<string | void> {
+  const output = await tryAwait(() => child.exec('solc --version', { env: SOLC_BIN_ENV }));
+  if (!output) return null;
+  const match = output.stdout.match(/^Version: ([^\s]+)/m);
+  if (!match) return null;
+  return match[1];
+}
+
+async function getAvailableCompilerVersions(): Promise<SolcList> {
+  const localPath = path.join(SOLC_CACHE_PATH, 'list.json');
+  mkdirpSync(SOLC_CACHE_PATH);
+  if (fs.existsSync(localPath)
+    && (Date.now() - (+fs.statSync(localPath).mtime)) < SOLC_LIST_EXPIRES_IN_SECONDS * 1000) {
+    return readJsonSync(localPath);
+  }
+
+  try {
+    const response = await axios.get(SOLC_LIST_URL);
+    const list = response.data as SolcList;
+    writeJsonSync(localPath, list);
+    return list;
+  } catch(err) {
+    if (fs.existsSync(localPath)) {
+      log.warn(`Error downloading solc releases list, using cached version`);
+      return readJsonSync(localPath);
+    } else {
+      throw new Error(`Could not retrieve solc releases list: ${err.message}`);
```

Comment:
`throw new Error(` erases stacktrace of a previous exception. We need to write utility function to rethrow exceptions and use it everywhere.